webapp/graphite/errors.py

```diff
@@ -7,7 +8,71 @@ class NormalizeEmptyResultError(Exception):
 
 
 class InputParameterError(ValueError):
-    pass
+
+    def __init__(self, *args, **kwargs):
+        super(InputParameterError, self).__init__(*args, **kwargs)
+        self.context = {}
```

Comment:
For this exception class to pickle+unpickle correctly, everything should be in args or kwargs, instead of using `setFoo()` methods to add more info.

... but maybe we can be sure that objects of this class will never be pickled, I'm not sure about that. If multiprocessing is used and an obj of this class ends up assigned to a local variable, that can happen. I've run into this problem before in other projects.

Reply:
(if relevant, also need to continue passing all args and kwargs to the parent class constructor in the same way as they were received, and then you probably want to override the `__str__()` method)

Reply:
> If multiprocessing is used and an obj of this class ends up assigned to a local variable, that can happen. I've run into this problem before in other projects.

That's interesting, i had no idea it might do that. This class should only be used within one request handling thread.

> For this exception class to pickle+unpickle correctly, everything should be in args or kwargs, instead of using setFoo() methods to add more info.

I'm not sure I understand that right. Why would pickle not also encode `InputParameterError.context` when that object gets encoded?

I tried that on the cli:

```
(Pdb) e.__class__
<class 'graphite.errors.InputParameterError'>
(Pdb) e.context
{'function': 'asPercent', 'args': [321], 'kwargs': {}, 'target': 'asPercent(321)', 'sourceIdHeaders': {}}
(Pdb) e.describe()
'Invalid parameters (Invalid "seriesList" value specified for function "asPercent" parameter "seriesList": 321); target: "asPercent(321)"; func: "asPercent(321)"'
(Pdb) restored = pickle.loads(pickle.dumps(e))
(Pdb) restored.context
{'function': 'asPercent', 'args': [321], 'kwargs': {}, 'target': 'asPercent(321)', 'sourceIdHeaders': {}}
(Pdb) restored.describe()
'Invalid parameters (Invalid "seriesList" value specified for function "asPercent" parameter "seriesList": 321); target: "asPercent(321)"; func: "asPercent(321)"'
````

Isn't that what we want it to do?

> and then you probably want to override the __str__() method

good idea

Reply:
Hmm yeah maybe this works fine.

In the case I previously saw, some arguments were assigned to other properties, and then *different* args were passed to the parent class constructor. It seemed that the Exception class handled pickle by saving the args, and for unpickle it re-created the subclass by passing the saved args to the sub-class constructor, and it was confused by receiving the *different* args it originally passed to its parent.

Your new exception already passes on the exact same args. I suspected context would end up empty again. I guess pickling preserves other object attributes of Exception classes too, and restores them after init, and the only real problem I had back when, was the changed args...

Reply:
Ok, thanks. So it looks like this is fine then, in that case I'll resolve this comment

Reply:
yup, all looks good to me